perses/dispersed/loopy_annealing_integrator.py

```diff
@@ -0,0 +1,153 @@
+#!/usr/bin/env python
+
+from openmmtools.integrators import AlchemicalNonequilibriumLangevinIntegrator
+from simtk import unit
+import numpy as np
+import simtk.openmm as openmm
+from openmmtools.constants import kB
+from openmmtools import utils
+
+
+# In[119]:
+
+
+class LoopyAlchemicalNonequilibriumLangevinIntegrator(AlchemicalNonequilibriumLangevinIntegrator):
+    """
+    Subclass of `AlchemicalNonequilibriumLangevinIntegrator` that hardcodes the following integration scheme:
+
+    Step 1: lambda_0 equilibrium sampling
+        run nsteps_eq of BAOAB (V R O R V) integration at the alchemical lambda_0 endstate
+    Step 2: Forward Annealing
+        run n_steps_neq of BAOAB (V R O R V H); 'H': Hamiltonian update step
+    Step 3: lambda_1 equilibrium sampling
+        run Step 1 at the lambda_1 endstate
+    Step 3: Reverse Annealing
+        run Step 2 in the reverse direction
+
+    forward and backward works are saved to `forward_work` and `backward_work`, respectively
+    """
+    def __init__(self,
+                 alchemical_functions = None,
+                 nsteps_eq = 1000,
+                 nsteps_neq = 100,
+                 **kwargs):
+        """        
+        arguments
+            nsteps_eq : int
+                number of equilibration steps to run at either endstate
+            nsteps_neq : int
+                number of nonequilibrium annealing steps to run in either direction
+
+        parameters
+            _n_steps_eq : int
+                number of BAOAB loops to run at each endstate in an attempt to generate i.i.d samples from endstates
+        """
+        self._n_steps_eq = nsteps_eq
+        splitting = 'V R O R V'
+        super().__init__(alchemical_functions, splitting, nsteps_neq = nsteps_neq, **kwargs)
+    
+    def _add_global_variables(self):
+        """
+        modify the super (_add_global_variables) to add a collector for forward and reverse annealing, as well as to specify the number of equilibration
+        steps at endstates.
+        """
+        super()._add_global_variables()
+        self.addGlobalVariable('forward_work', 0)
```

Comment:
@dominicrufa : We don't want to separately accumulate forward and backward work, since it will get very complicated to deconvolute. We just want to accumulate the total `protocol_work` and then extract differences in the total work at different times to get the work for each switch.

Reply:
Or do we just _also_ separate the work into forward and backward, in _addition_ to accumulating `protocol_work`? If so, I don't see any problem with doing both!

Reply:
I'm confused about what you mean here; can you explain further?

Reply:
you don't want to accumulate forward and backward works for each step and then use a `BAR` estimator?

Reply:
i _think_ i understand what you are talking about, but let me be clear and you can raise any concerns.  at the start of an integration step, the `forward_work`, `backward_work` and `protocol_work` are all zeroed.  the `protocol_work` is accumulated in the _forward_ annealing, passed to `forward_work`, and zeroed.  `protocol_work` is re-accumulated (from zero) in the backward direction, and then passed to `backward_work`

Reply:
strictly speaking, there is no reason to create a `backward_work` variable, since it is contained in the `protocol_work` at the end of the entire integration step, but since i created a `forward_work` variable to log the accumulated work in the forward pass, it made sense to me to also create a `backward_work` variable